src/sentry/interfaces/http.py

```diff
@@ -167,13 +167,28 @@ def to_python(cls, data):
         else:
             headers = ()
 
+        # We prefer the body to be a string, since we can then attempt to parse it
+        # as JSON OR decode it as a URL encoded query string, without relying on
+        # the correct content type header being passed.
         body = data.get('data')
-        if isinstance(body, dict):
-            body = json.dumps(body)
+
+        content_type = next((v for k, v in headers if k == 'Content-Type'), None)
+
+        # Remove content type parameters
+        if content_type is not None:
+            content_type = content_type.partition(';')[0].rstrip()
+
+        # This may be the second time we're processing our request data. The
+        # inferred_content_type may have already been set.
```

Comment:
This is definitely helpful, but a little more elaboration on why this could be processed again would be even more helpful.

Reply:
Also, doesn't this mean that content detection and decoding the `data` content could be entirely behind a `data.get('inferred_content_type') is not None` check to avoid doing the work again after it's already been done on the first pass?

Reply:
I didn't investigate too deeply to understand why we do this again when requesting the details over the API, so I don't want to make any assumptions in this comment. I'll at least mention that it happens during store and then again at request time.

> Also, doesn't this mean that content detection and decoding the `data` content could be entirely behind a `data.get('inferred_content_type') is not None` check to avoid doing the work again after it's already been done on the first pass?

Yes, but the `if not isinstance(body, dict):` condition is basically already doing that, since it will have been decoded into a body before. Though I suppose we could avoid *not being able to decode* twice by doing that check. Sounds good.

Reply:
f20adb1da7ddd2e3d9776f08367e32593f2ad897